program/rust/src/tests/pyth_simulator.rs

```diff
@@ -67,10 +67,49 @@ use {
         fs::File,
         iter::once,
         mem::size_of,
-        path::Path,
+        path::PathBuf,
     },
 };
 
+lazy_static::lazy_static! {
+    // Build the oracle binary and make it available to the
+    // simulator. lazy_static makes this happen only once per test
+    // run.
+    static ref ORACLE_PROGRAM_BINARY_PATH: PathBuf = {
+
+    // Detect features and pass them onto cargo-build-bpf.
+    // IMPORTANT: All features of this crate must have gates added to this vector.
+    let features: Vec<&str> = vec![
+    #[cfg(feature = "pythnet")]
+    "pythnet",
+
+    #[cfg(feature = "price_v2_resize")]
+    "price_v2_resize",
+
+    ];
+
+    let mut cmd = std::process::Command::new("cargo");
+    cmd.arg("build-bpf");
+
+    if !features.is_empty() {
+        cmd.arg("--features");
+        cmd.args(features);
+    }
+
+    let status = cmd.status().unwrap();
+
+    if !status.success() {
+        panic!(
+        "cargo-build-bpf did not exit with 0 (code {:?})",
+        status.code()
+        );
+    }
+
+    let target_dir = concat!(env!("CARGO_MANIFEST_DIR"), "/../../target");
+
+    PathBuf::from(target_dir).join("deploy/pyth_oracle.so")
+    };
```

Comment:
The root problem is that `build.rs` has no way of knowing that it is being used for a `cargo test` build (this is deliberate as pointed [here](https://github.com/rust-lang/cargo/issues/4789). People tend to solve it with dedicated features (which would get passed to `build.rs`). Our workaround was just triggering a `build-bpf` every time we build for a non-`bpf` target arch. This would work for `cargo test`, but it would also build BPF on `cargo build`, `cargo check` which absolutely do not need it.

My specific reason for moving this is pre commit checks. Our `pre-commit` setup indirectly does `cargo check` which still will evaluate the build.rs - meaning a working Solana toolchain would be necessary for running pre-commit if we kept the bpf build there. Incidentally, CI worked up to this point due to a separate problem - we didn't check the `cargo-build-bpf` exit code inside the `build.rs`, so a misconfigured toolchain would go unnoticed by pre-commit. While it means that I could've left this as is, the unchecked BPF build result could likely cause very strange `cargo test` behaviors (imagine you built a BPF binary but made a change that would silently mess the BPF build up - your tests would be ran against the pre-existing binary).